File: reference/java/src/main/java/com/yawlabs/awsp/InMemoryReplayStore.java
// AWSP -- A2A Webhook Security Profile reference implementation.
// Copyright 2026 Yaw Labs. Licensed under the Apache License, Version 2.0.

package com.yawlabs.awsp;

import java.nio.charset.StandardCharsets;
import java.time.Clock;
import java.util.Iterator;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicLong;
import java.util.function.LongSupplier;

/**
 * Reference {@link ReplayStore} backed by an in-memory map. Suitable for
 * tests and single-process receivers; production multi-replica deployments
 * should use Redis ({@code SET key NX EX}) or equivalent so replay state is
 * shared across nodes.
 *
 * <p>This implementation is thread-safe (via {@link ConcurrentHashMap}). Its
 * eviction sweep is best-effort: above {@link #EVICT_THRESHOLD} entries the
 * sweep runs at most once every {@link #EVICT_SWEEP_INTERVAL} calls (so the
 * amortized per-call cost stays O(1) instead of O(n) once the map fills).
 *
 * <p>Bound on map size: at steady state the map holds at most
 * {@code peak_arrival_rate * (ttlSeconds + EVICT_SWEEP_INTERVAL)} entries --
 * the sweep window adds at most {@code EVICT_SWEEP_INTERVAL} extra entries
 * past the natural TTL bound before eviction reclaims them.
 */
public final class InMemoryReplayStore implements ReplayStore {

    /** Map size at or above which the eviction sweep is considered. */
    private static final int EVICT_THRESHOLD = 8192;

    /**
     * Once above {@link #EVICT_THRESHOLD}, run the O(n) sweep at most once
     * per this many calls. Cheap counter-modulo amortizes the sweep cost
     * across the call stream.
     */
    private static final int EVICT_SWEEP_INTERVAL = 256;

    private final ConcurrentHashMap<String, Long> seen = new ConcurrentHashMap<>();
    private final LongSupplier clock;
    private final AtomicLong callsSinceSweep = new AtomicLong();

    /** Construct with the system clock (unix-seconds resolution). */
    public InMemoryReplayStore() {
        this(() -> System.currentTimeMillis() / 1000L);
    }

    /** Construct with a custom clock. Useful for deterministic tests. */
    public InMemoryReplayStore(LongSupplier clock) {
        this.clock = clock;
    }

    /** Construct from a {@link Clock} instance. */
    public InMemoryReplayStore(Clock javaClock) {
        this(() -> javaClock.instant().getEpochSecond());
    }

    @Override
    public boolean checkAndStore(String configId, byte[] nonce, int ttlSeconds) {
        long now = clock.getAsLong();
        evict(now);
        String key = makeKey(configId, nonce);
        long expiresAt = now + ttlSeconds;
        // putIfAbsent semantics: returns null if the key was unseen and
        // we successfully claimed it. Otherwise returns the existing value
        // -- we then check whether it has already expired (in which case
        // we replace and accept the nonce as fresh).
        Long prior = seen.putIfAbsent(key, expiresAt);
        if (prior == null) {
            return true;
        }
        if (prior <= now) {
            // Stale entry left behind by a slow eviction sweep. Race-safely
            // try to replace it with the fresher expiry.
            if (seen.replace(key, prior, expiresAt)) {
                return true;
            }
            // Lost the race -- some other caller already inserted a fresh
            // entry. That's a replay, fall through.
        }
        return false;
    }

    /** Number of nonces currently tracked (including expired-but-not-swept). */
    public int size() {
        return seen.size();
    }

    private static String makeKey(String configId, byte[] nonce) {
        // Nonces are base64url ASCII per the spec; UTF-8 round-trips them
        // unambiguously. Concatenating with a separator that cannot appear
        // in either field keeps configId-vs-nonce ambiguity from arising.
        // (configId is opaque, but ":" is the conventional split char.)
        String n = new String(nonce, StandardCharsets.UTF_8);
        return configId + ":" + n;
    }

    private void evict(long now) {
        if (seen.size() < EVICT_THRESHOLD) {
            // Map is small. Reset the counter so the first sweep above the
            // threshold runs immediately rather than after the modulo cycle.
            callsSinceSweep.set(0);
            return;
        }
        // O(n) sweep -- only runs once per EVICT_SWEEP_INTERVAL calls past
        // the threshold, so amortized cost per call stays O(1). Memory is
        // bounded by arrival rate x (ttl + EVICT_SWEEP_INTERVAL); for
        // higher-throughput needs, wrap a TTL-aware data structure
        // (Caffeine, etc.) instead.
        long n = callsSinceSweep.getAndIncrement();
        if (n % EVICT_SWEEP_INTERVAL != 0) {
            return;
        }
        Iterator<Map.Entry<String, Long>> it = seen.entrySet().iterator();
        while (it.hasNext()) {
            Map.Entry<String, Long> e = it.next();
            if (e.getValue() <= now) {
                it.remove();
            }
        }
    }
}

File: reference/dotnet/src/YawLabs.Awsp/InMemoryReplayStore.cs
// AWSP -- A2A Webhook Security Profile reference implementation.

using System;
using System.Collections.Generic;

namespace YawLabs.Awsp;

/// <summary>
/// Reference in-memory <see cref="IReplayStore"/>. Suitable for tests and single-process
/// receivers; production multi-replica deployments should use Redis (<c>SET NX EX</c>) or
/// equivalent so replay state is shared.
/// <para>
/// All public methods are thread-safe. Eviction is incremental and runs only when the store
/// grows past a threshold; for higher-throughput workloads, use a TTL-aware backing store.
/// </para>
/// </summary>
public sealed class InMemoryReplayStore : IReplayStore
{
    // Eviction strategy: once the map crosses EvictionThreshold entries, sweep only every
    // EvictionInterval-th call. Sweeping on every call past the threshold made the worst-case
    // CheckAndStore O(n) per call -- a pathological pattern of monotonically-fresh nonces with
    // no expirations made the whole store quadratic in steady state. The interval-throttled
    // sweep keeps the amortized cost O(1) while still bounding map growth: between sweeps the
    // map can grow by at most EvictionInterval entries, so the steady-state ceiling is
    // EvictionThreshold + EvictionInterval.
    private const int EvictionThreshold = 8192;
    private const int EvictionInterval = 256;

    private readonly object _lock = new();
    private readonly Dictionary<string, long> _seen = new(StringComparer.Ordinal);
    private readonly Func<DateTimeOffset> _clock;
    private long _callsSinceLastSweep;

    /// <summary>
    /// Create a store using <see cref="DateTimeOffset.UtcNow"/> as the clock.
    /// </summary>
    public InMemoryReplayStore()
        : this(static () => DateTimeOffset.UtcNow)
    {
    }

    /// <summary>
    /// Create a store with an injectable clock. Useful for deterministic tests.
    /// </summary>
    public InMemoryReplayStore(Func<DateTimeOffset> clock)
    {
        ArgumentNullException.ThrowIfNull(clock);
        _clock = clock;
    }

    /// <inheritdoc />
    public bool CheckAndStore(string configId, byte[] nonce, int ttlSeconds)
    {
        ArgumentNullException.ThrowIfNull(nonce);
        if (ttlSeconds <= 0)
        {
            throw new ArgumentOutOfRangeException(nameof(ttlSeconds), ttlSeconds, "ttlSeconds must be positive");
        }

        // Nonce uniqueness is global to the receiver per AWSP section 7.2; the kid/configId is
        // not part of the dedupe key. configId is allowed only as a multi-tenant prefix when
        // useful; the reference implementation prefixes it for parity with Redis-style stores.
        string key = MakeKey(configId, nonce);
        long now = _clock().ToUnixTimeSeconds();

        lock (_lock)
        {
            EvictIfLarge(now);
            if (_seen.TryGetValue(key, out long expiresAt) && expiresAt > now)
            {
                return false;
            }
            _seen[key] = now + ttlSeconds;
            return true;
        }
    }

    private static string MakeKey(string configId, byte[] nonce)
    {
        // Convert(byte[]) emits standard base64; sufficient for an internal map key. We do NOT
        // round-trip this back to a base64url nonce -- the dictionary just needs a stable
        // string representation per byte sequence.
        string nonceKey = Convert.ToBase64String(nonce);
        return string.IsNullOrEmpty(configId) ? nonceKey : (configId + ":" + nonceKey);
    }

    private void EvictIfLarge(long now)
    {
        if (_seen.Count <= EvictionThreshold)
        {
            // Reset the call counter so a future shrink-then-grow cycle does not sweep on the
            // very first call past the threshold.
            _callsSinceLastSweep = 0;
            return;
        }
        // Throttled sweep: only walk the map every EvictionInterval-th call past threshold.
        // This keeps amortized CheckAndStore cost O(1) instead of O(n) under sustained load.
        _callsSinceLastSweep++;
        if (_callsSinceLastSweep < EvictionInterval)
        {
            return;
        }
        _callsSinceLastSweep = 0;

        // O(n) sweep -- bounded because n is also bounded by the time window.
        var expired = new List<string>();
        foreach (var pair in _seen)
        {
            if (pair.Value <= now)
            {
                expired.Add(pair.Key);
            }
        }
        foreach (string k in expired)
        {
            _seen.Remove(k);
        }
    }
}
